virtcontainers/cgroups.go

```diff
@@ -1,197 +1,251 @@
-	if s.cgroup == nil {
-		return fmt.Errorf("failed to setup uninitialized cgroup for sandbox")
+func (s *Sandbox) updateCgroups() error {
+	if s.state.CgroupPath == "" {
+		return fmt.Errorf("Cgroup path is empty")
 	}
 
-	resource, err := s.mergeSpecResource()
+	cgroup, err := cgroups.Load(V1Constraints, cgroups.StaticPath(s.state.CgroupPath))
 	if err != nil {
+		return fmt.Errorf("Could not load cgroup %v: %v", s.state.CgroupPath, err)
+	}
+
+	if err := s.constrainHypervisor(cgroup); err != nil {
 		return err
 	}
 
-	if err := s.applyCPUCgroup(resource); err != nil {
+	if len(s.containers) <= 1 {
+		// nothing to update
+		return nil
+	}
+
+	resources, err := s.resources()
+	if err != nil {
 		return err
 	}
+
+	if err := cgroup.Update(&resources); err != nil {
+		return fmt.Errorf("Could not update cgroup %v: %v", s.state.CgroupPath, err)
+	}
+
 	return nil
 }
 
-func (s *Sandbox) applyCPUCgroup(rc *specs.LinuxResources) error {
-	if s.cgroup == nil {
-		return fmt.Errorf("failed to setup uninitialized cgroup for sandbox")
+func (s *Sandbox) deleteCgroups() error {
+	path := cgroupNoConstraintsPath(s.state.CgroupPath)
+	noConstraintsCgroup, err := cgroups.Load(V1NoConstraints, cgroups.StaticPath(path))
+	if err != nil {
+		return fmt.Errorf("Could not load cgroup %v: %v", path, err)
 	}
 
-	// apply cpu constraint to vcpu cgroup
-	if err := s.cgroup.vcpuSub.Update(rc); err != nil {
-		return err
+	return noConstraintsCgroup.Delete()
+}
+
+func (s *Sandbox) constrainHypervisor(cgroup cgroups.Cgroup) error {
+	pid := s.hypervisor.pid()
+	if pid <= 0 {
+		return fmt.Errorf("Invalid hypervisor PID: %d", pid)
 	}
 
-	// when new container joins, new CPU could be hotplugged, so we
-	// have to query fresh vcpu info from hypervisor for every time.
-	tids, err := s.hypervisor.getThreadIDs()
+	resources := &specs.LinuxResources{}
+	path := cgroupNoConstraintsPath(s.state.CgroupPath)
+	noConstraintsCgroup, err := cgroups.New(V1NoConstraints, cgroups.StaticPath(path), resources)
 	if err != nil {
-		return fmt.Errorf("failed to get thread ids from hypervisor: %v", err)
-	}
-	if tids == nil {
-		// If there's no tid returned from the hypervisor, this is not
-		// a bug. It simply means there is nothing to constrain, hence
-		// let's return without any error from here.
-		return nil
+		return fmt.Errorf("Could not create cgroup %v: %v", path, err)
 	}
 
-	// use Add() to add vcpu thread to s.cgroup, it will write thread id to
-	// `cgroup.procs` which will move all threads in qemu process to this cgroup
-	// immediately as default behaviour.
-	if len(tids.vcpus) > 0 {
-		if err := s.cgroup.sandboxSub.Add(cgroups.Process{
-			Pid: tids.vcpus[0],
-		}); err != nil {
-			return err
-		}
+	if err := noConstraintsCgroup.Add(cgroups.Process{Pid: pid}); err != nil {
+		return fmt.Errorf("Could not add hypervisor PID %d to cgroup %v: %v", pid, path, err)
 	}
 
-	for _, i := range tids.vcpus {
-		if i <= 0 {
-			continue
-		}
```

Comment:
Adding only the `vcpus` threads is an optimization, which can benefit to small chunks IO performance according to our previous test.

It's OK to remove this if it makes the handling logic too complicated, I'm just wishing to share the performance tuning  result to community.

Reply:
sure I can check that, what tests did you run?

Reply:
You can test with some network benchmark tools such as netperf, or disk IO benchmark tools such as FIO or iozone.  Try to send/receive with small chunks(e.g. 4k package), for one VM with 1 cpu, you can compare W/ or W/O "cpu cgroup of 1 cpu constraint", I believe you can see how the performance is different.

Reply:
you're right, performance is not good moving the whole hypervisor, I have updated the pull request, now the performance is good, thanks

![cgroup1](https://user-images.githubusercontent.com/13009432/52752844-ed684680-2fb9-11e9-9c44-26827bbcfee9.png)

![cgroup2](https://user-images.githubusercontent.com/13009432/52752852-f0fbcd80-2fb9-11e9-8338-ffbd4bed7a0e.png)


current - current implementation
new - new implementation
no-cgroup - container without constraints